File: assignment5/wumpus.zig
//PROGRAMMERS: Joao Engel, Teran Upchurch, Brendon Amino
//DATE: 10/08/24
//FILE: wumpus.zig
//
// Hunt the Wumpus Game
//
// A text-based game where players navigate a 6x4 maze to hunt the Wumpus.
// Players can move or shoot to defeat the Wumpus while avoiding deadly pits.
// Sensory cues indicate nearby dangers, and the game features random placement
// of the Wumpus and pits for varied experiences.


const std = @import("std");
const print = std.debug.print;
const expect = std.testing.expect;


// Define a 4x6 maze where each cell starts as an 'O' (open space)
var maze = [4][6]u8{
   .{ 'O', 'O', 'O', 'O', 'O', 'O' },
   .{ 'O', 'O', 'O', 'O', 'O', 'O' },
   .{ 'O', 'O', 'O', 'O', 'O', 'O' },
   .{ 'O', 'O', 'O', 'O', 'O', 'O' },
};


// Player's initial position
var playerRow: usize = 3;
var playerCol: usize = 0;


// Function to print the contents of the maze
fn printMaze() void {
   for (0..4) |i| {
       print("{s}\n", .{maze[i]});
   }
}


// Prints the map and player's current location on it with an 'X'
fn printMap(curPlayerRow: usize, curPlayerCol: usize) void {
   for (0..4) |r| {
       for (0..6) |c| {
           if (r == curPlayerRow and c == curPlayerCol) {
               print("X", .{}); // Player's current position
           } else {
               print("O", .{}); // Open space
           }
       }
       print("\n", .{}); // Move to the next line after each row
   }
}


// Initialize the maze by randomly placing pits and a wumpus
fn initMaze() void {
   const rand = std.crypto.random; // Random number generator
   var col: usize = 0;
   var row: usize = 0;


   // Place four pits (one per row)
   for (0..4) |r| {
       col = rand.intRangeAtMost(u8, 1, 5); // Random column for pit
       maze[r][col] = 'P'; // Place pit
   }


   // Place the wumpus
   col = rand.intRangeAtMost(u8, 1, 5); // Random column for wumpus
   row = rand.intRangeAtMost(u8, 0, 3); // Random row for wumpus
   while (maze[row][col] == 'P') { // Ensure wumpus is not placed in a pit
       col = rand.intRangeAtMost(u8, 1, 5);
       row = rand.intRangeAtMost(u8, 0, 3);
   }


   maze[row][col] = 'W'; // Place wumpus
}


// Function to move the player based on input
fn characterMove(playerMove: u8, curPlayerRow: usize, curPlayerCol: usize) void {
   if (playerMove == 'u' and playerRow > 0 and checkWall(curPlayerRow - 1, curPlayerCol)) {
       playerRow -= 1; // Move up
       checkMove(playerRow, playerCol); // Check new position
   } else if (playerMove == 'd' and checkWall(curPlayerRow + 1, curPlayerCol)) {
       playerRow += 1; // Move down
       checkMove(playerRow, playerCol);
   } else if (playerMove == 'l' and playerCol > 0 and checkWall(curPlayerRow, curPlayerCol - 1)) {
       playerCol -= 1; // Move left
       checkMove(playerRow, playerCol);
   } else if (playerMove == 'r' and checkWall(curPlayerRow, curPlayerCol + 1)) {
       playerCol += 1; // Move right
       checkMove(playerRow, playerCol);
   } else {
       print("\nOuch! You ran into a wall\n", .{}); // Collision with wall
   }
}


// Check if the new position is within maze boundaries
fn checkWall(newPlayerRow: usize, newPlayerCol: usize) bool {
   if ((newPlayerRow > 3) or (newPlayerRow < 0) or (newPlayerCol > 5) or (newPlayerCol < 0)) {
       return false; // Out of bounds, invalid move
   }
   return true; // Valid move
}


// Check for wumpus or pit in the player's current position
fn checkMove(newRow: usize, newCol: usize) void {
   // Check if the player found the wumpus
   if (maze[newRow][newCol] == 'W') {
       print("\nYou've been eaten by the WUMPUS!\n", .{});
       printMaze(); // Show the maze before exiting
       std.process.exit(0); // Exit the game
   }
   // Check if the player fell into a pit
   else if (maze[newRow][newCol] == 'P') {
       print("\nYou've fallen into a pit and have perished!\n", .{});
       printMaze(); // Show the maze before exiting
       std.process.exit(0); // Exit the game
   }


   // Check for proximity to pits or wumpus for warning messages
   // Check the cell above
   if (newRow > 0 and maze[newRow - 1][newCol] != 'O') {
       if (maze[newRow - 1][newCol] == 'P') {
           print("You feel a breeze...\n", .{}); // Nearby pit
       } else {
           print("You smell a stench...\n", .{}); // Nearby wumpus
       }
   }
   // Check the cell below
   if (newRow < 3 and maze[newRow + 1][newCol] != 'O') {
       if (maze[newRow + 1][newCol] == 'P') {
           print("You feel a breeze...\n", .{}); // Nearby pit
       } else {
           print("You smell a stench...\n", .{}); // Nearby wumpus
       }
   }
   // Check the cell to the right
   if (newCol < 5 and maze[newRow][newCol + 1] != 'O') {
       if (maze[newRow][newCol + 1] == 'P') {
           print("You feel a breeze...\n", .{}); // Nearby pit
       } else {
           print("You smell a stench...\n", .{}); // Nearby wumpus
       }
   }
   // Check the cell to the left
   if (newCol > 0 and maze[newRow][newCol - 1] != 'O') {
       if (maze[newRow][newCol - 1] == 'P') {
           print("You feel a breeze...\n", .{}); // Nearby pit
       } else {
           print("You smell a stench...\n", .{}); // Nearby wumpus
       }
   }
}


// Check if the wumpus is in the adjacent space in the direction the player is shooting
fn checkShoot(dir: u8, curPlayerRow: usize, curPlayerCol: usize) bool {
   if (dir == 'u') {
       if (curPlayerRow != 0 and maze[curPlayerRow - 1][curPlayerCol] == 'W') {
           return true; // Wumpus is above
       }
   } else if (dir == 'r') {
       if (curPlayerCol != 5 and maze[curPlayerRow][curPlayerCol + 1] == 'W') {
           return true; // Wumpus is to the right
       }
   } else if (dir == 'd') {
       if (curPlayerRow != 3 and maze[curPlayerRow + 1][curPlayerCol] == 'W') {
           return true; // Wumpus is below
       }
   } else if (dir == 'l') {
       if (curPlayerCol != 0 and maze[curPlayerRow][curPlayerCol - 1] == 'W') {
           return true; // Wumpus is to the left
       }
   }
   return false; // Wumpus is not in the shooting direction
}


// Main game function
pub fn main() !void {
   const reader = std.io.getStdIn().reader(); // Standard input reader
   var buffer: [16]u8 = undefined; // Buffer for user input
   var choice: u8 = '1'; // User's action choice
   var shotChoice: u8 = '1'; // Direction to shoot


   initMaze(); // Initialize maze with pits and wumpus
   printMap(playerRow, playerCol); // Display initial player position
   checkMove(playerRow, playerCol); // Check initial surroundings


   // Game loop
   while (choice != 'q') {
       // Display options to the player
       print("What do you want to do:\n u) move up\n", .{});
       print(" d) move down\n l) move left\n r) move right\n s) shoot arrow\n q) quit\n ENTER CHOICE: ", .{});
       const input = try reader.readUntilDelimiter(&buffer, '\n');
       choice = input[0]; // Get user choice
       if (choice == 's') {
           print("\nWhat direction do you want to shoot your arrow:\n", .{});
           print(" u) up\n d) down\n l) left\n r) right\n ENTER CHOICE: ", .{});
           const direction = try reader.readUntilDelimiter(&buffer, '\n');
           shotChoice = direction[0]; // Get shooting direction
           // Validate shooting direction
           if (shotChoice != 'u' and shotChoice != 'd' and shotChoice != 'l' and shotChoice != 'r') {
               print("\nInvalid choice\n\n", .{});
               printMap(playerRow, playerCol);
               continue;
           }
           // Check if shot hit the wumpus
           if (!checkShoot(shotChoice, playerRow, playerCol)) {
               print("You missed and failed to slay the WUMPUS!\n", .{});
           } else {
               print("You slayed the WUMPUS! You win!\n", .{});
           }
           printMaze(); // Show the maze after shooting
           break; // End the game after shooting
       } else if (choice == 'u') {
           characterMove(choice, playerRow, playerCol); // Move up
           printMap(playerRow, playerCol);
       } else if (choice == 'd') {
           characterMove(choice, playerRow, playerCol); // Move down
           printMap(playerRow, playerCol);
       } else if (choice == 'l') {
           characterMove(choice, playerRow, playerCol); // Move left
           printMap(playerRow, playerCol);
       } else if (choice == 'r') {
           characterMove(choice, playerRow, playerCol); // Move right
           printMap(playerRow, playerCol);
       } else if (choice != 'q') {
           print("\nInvalid choice\n\n", .{}); // Handle invalid choices
       }
   }
}

File: assignment6/wumpus.rb
#PROGRAMMERS: Joao Engel, Brendon Amino, Teran Upchurch
#DATE: 10/09/24
#FILE: wumpus.rb

# A text-based game where players navigate a 6x4 maze to hunt the Wumpus.
# Players can move or shoot to defeat the Wumpus while avoiding deadly pits.
# Sensory cues indicate nearby dangers, and the game features random placement
# of the Wumpus and pits for varied experiences.


# Base class Tile represents a generic tile in the maze, storing sensory queue for nearby dangers
class Tile
    attr_accessor :sensoryQueue
 
    # Initializes an empty sensoryQueue string
    def initialize
        @sensoryQueue = ""
    end
  
    # Sets the sensoryQueue with the provided action string
    def setSensoryQueue(action)
        @sensoryQueue = action
    end
 
    # Retrieves the sensoryQueue
    def getSensoryQueue
        @sensoryQueue
    end

    def visit #By default returns newline, used to print tile specific messages
        puts ""
    end
 end
 
 
 # Derived class Location (formerly EmptyTile) represents an empty tile in the maze
 class Location < Tile
    def type #In place of "print"
        "O"  # Denotes empty tile type
    end
 end
 
 
 # Derived class Pit represents a pit in the maze
 class Pit < Tile
    def visit #Override
        puts "You've fallen into a pit and have perished!"
    end
    
    def type
        "P"  # Denotes pit tile type
    end
 end
 

 # Derived class Wumpus represents the Wumpus in the maze
 class Wumpus < Tile
    def visit #Override
        puts "You've been eaten by the WUMPUS!"
    end

    def type
        "W"  # Denotes Wumpus tile type
    end
 end
 
 
 # The Game class manages the game board and player's actions
 class Game
    attr_accessor :maze, :playerRow, :playerCol

    # Initializes a 6x4 empty maze and sets the player start position
    def initialize()
        @maze = [
        [Location.new, Location.new, Location.new, Location.new, Location.new, Location.new],
        [Location.new, Location.new, Location.new, Location.new, Location.new, Location.new],
        [Location.new, Location.new, Location.new, Location.new, Location.new, Location.new],
        [Location.new, Location.new, Location.new, Location.new, Location.new, Location.new]
        ]
        @playerRow = 3  # Player starts at the bottom left
        @playerCol = 0
    end

    # Randomly places four pits and the Wumpus in the maze
    def initMaze
        # Place four pits (one per row, randomly in columns 1-5)
        for r in 0..3
            col = rand(1..5)
            @maze[r][col] = Pit.new
            addActionsToTile(r, col, "P")  # Adds sensory cues for nearby pits
        end
 
        # Place the Wumpus in a random tile, avoiding pits
        row = rand(0..3)
        col = rand(1..5)
        while @maze[row][col].type == "P"
            row = rand(0..3)
            col = rand(1..5)
        end
        @maze[row][col] = Wumpus.new
        addActionsToTile(row, col, "W")  # Adds sensory cues for nearby Wumpus
    end

    # Prints the maze showing each tile type (P = Pit, W = Wumpus, O = Empty)
    def printMaze
        for r in 0..3 do
            for c in 0..5 do
                print "#{@maze[r][c].type} "
            end
            @maze[3][0].visit #Player starting location: always considered empty tile and will thus return newline
        end
    end
 
    # Prints the player's current map, showing "X" for the player's position
    def printMap
        for r in 0..3 do
            for c in 0..5 do
                if r == @playerRow and c == @playerCol
                    print "X "  # Player position
                else
                    print "O "  # Empty tile
                end
            end
            puts ""
        end
    end
 
    # Moves the player based on input (u = up, d = down, l = left, r = right)
    def playerMove(letter)
        if letter == 'u' and @playerRow > 0
            @playerRow -= 1  # Move up
        elsif letter == 'd' and @playerRow < 3
            @playerRow += 1  # Move down
        elsif letter == 'l' and @playerCol > 0
            @playerCol -= 1  # Move left
        elsif letter == 'r' and @playerCol < 5
            @playerCol += 1  # Move right
        elsif letter == 'q'
            print ""
        else
            print "Ouch! You ran into a wall\n"  # Collision with wall
        end
        checkMove
    end
 
    # Checks the current tile for hazards (Wumpus or Pit) after each move
    def checkMove
        playerCurrentPos = @maze[@playerRow][@playerCol]
        if playerCurrentPos.type == "W" #Checks for Wumpus, then calls Wumpus tile's visit method
            playerCurrentPos.visit
            printMaze
            exit(0)  # End game
        elsif playerCurrentPos.type == "P" #Checks for Pit, then calls Pit tile's visit method
            playerCurrentPos.visit
            printMaze
            exit(0)  # End game
        end
 
        # Provide sensory feedback about nearby dangers
        print "#{playerCurrentPos.getSensoryQueue}"
    end
 
    # Shoots an arrow in the specified direction to try to slay the Wumpus
    def playerShoot(choice)
        if choice == 'u' and @playerRow > 0 and @maze[@playerRow - 1][@playerCol].type == "W"
            puts "\nYou slayed the WUMPUS! You win!"
        elsif choice == 'd' and @playerRow < 3 and @maze[@playerRow + 1][@playerCol].type == "W"
            puts "\nYou slayed the WUMPUS! You win!"
        elsif choice == 'l' and @playerCol > 0 and @maze[@playerRow][@playerCol - 1].type == "W"
            puts "\nYou slayed the WUMPUS! You win!"
        elsif choice == 'r' and @playerCol < 5 and @maze[@playerRow][@playerCol + 1].type == "W"
            puts "\nYou slayed the WUMPUS! You win!"
        else
            puts "\nYou missed and failed to slay the WUMPUS!\n"
        end
        printMaze
        exit(0)
    end
 
    # Adds sensory cues ("breeze" for pits, "stench" for Wumpus) to adjacent tiles
    def addActionsToTile(row, col, type)
        action_text = type == "P" ? "You feel a breeze...\n" : "You smell a stench...\n"
      
        # Add sensory cues to adjacent tiles (up, down, left, right)
        if row > 0 and @maze[row - 1][col].type == "O"
            @maze[row - 1][col].setSensoryQueue(@maze[row - 1][col].getSensoryQueue + action_text)
        end
        if row < 3 and @maze[row + 1][col].type == "O"
            @maze[row + 1][col].setSensoryQueue(@maze[row + 1][col].getSensoryQueue + action_text)
        end
        if col > 0 and @maze[row][col - 1].type == "O"
            @maze[row][col - 1].setSensoryQueue(@maze[row][col - 1].getSensoryQueue + action_text)
        end
        if col < 5 and @maze[row][col + 1].type == "O"
            @maze[row][col + 1].setSensoryQueue(@maze[row][col + 1].getSensoryQueue + action_text)
        end
    end
 
    # Main game loop where players can move, shoot, or quit
    def play
        choice = "O"
        checkMove # letting player know if it is next to a pit or wumpus before first move
        while choice != "q"
            puts ""
            puts "What do you want to do:"
            puts " u) Move Up"
            puts " d) Move Down"
            puts " l) Move Left"
            puts " r) Move Right"
            puts " s) Shoot Arrow"
            puts " q) Quit"
            print "ENTER CHOICE: "
            choice = gets.chomp
            puts ""

            if choice == "q" #If player quits, prints quitting game and doesn't print anything else
                puts "Quitting game..."
                #printMaze
                break
            end
            if choice != "s" and choice != "u" and choice != "d" and choice != "l" and choice != "r" and choice != "q"
                puts "Invalid choice!"
                printMap
            elsif choice == "s"
                puts "What direction do you want to shoot your arrow:"
                puts " u) Shoot Up"
                puts " d) Shoot Down"
                puts " l) Shoot Left"
                puts " r) Shoot Right"
                print "ENTER CHOICE: "
                shotDirection = gets.chomp
 
                #If shot direction isn't a valid letter,
            if shotDirection != "u" and shotDirection != "d" and shotDirection != "l" and shotDirection != "r"
                puts ""
                puts "Invalid choice!"
                printMap
            else
                playerShoot(shotDirection)
            end
            else
                playerMove(choice)
                printMap  # Print the player's map showing position after move
            end
        end
    end
 end


 # main program execution
 c = Game.new #create a new Game object called c
 c.printMap
 c.initMaze
 c.play  # starts the game menu